### PR TITLE
Adding OpenJDK 11 from SapMachine

### PR DESCRIPTION
**Pull request by gjmveloso**
After making all changes to the cask:

- [x] `brew cask audit --download {{cask_file}}` is error-free.
- [x] `brew cask style --fix {{cask_file}}` reports no offenses.
- [x] The commit message includes the cask’s name and version.
- [x] The submission is for [a stable version](https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#finding-a-home-for-your-cask) or [documented exception](https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#but-there-is-no-stable-version).

Additionally, if **adding a new cask**:

- [x] Named the cask according to the [token reference].
- [x] `brew cask install {{cask_file}}` worked successfully.
- [x] `brew cask uninstall {{cask_file}}` worked successfully.
- [x] Checked there are no [open pull requests] for the same cask.
- [x] Checked the cask was not [already refused].
- [x] Checked the cask is submitted to [the correct repo].

[token reference]: https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/token_reference.md
[open pull requests]: https://github.com/Homebrew/homebrew-cask/pulls
[already refused]: https://github.com/Homebrew/homebrew-cask/search?q=is%3Aclosed&type=Issues
[the correct repo]: https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#finding-a-home-for-your-cask
[version-checksum]: https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/sha256.md#updating-the-sha256

### CODE REVIEW

**Comment by gjmveloso** [comment]
Did this per discussion at [homebrew-core/issues/30872](https://github.com/Homebrew/homebrew-core/issues/30872).
Java becomes available after a shell restart. Might need to add other binaries like `javac` and others.

File: Casks/openjdk.rb [hunk 1]
```diff
@@ -10,5 +10,24 @@
   auto_updates true
 
   binary "/Library/Java/JavaVirtualMachines/#{version}/bin/java"
+  binary "/Library/Java/JavaVirtualMachines/#{version}/bin/javac"
   artifact "#{version}/", target: "/Library/Java/JavaVirtualMachines/#{version}"
+
+  uninstall launchctl: [
```
Comment (commitay) [inline Casks/openjdk.rb]: Remove all of the `uninstall`, it isn't needed with this cask.

File: Casks/openjdk.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
+cask 'openjdk' do
+  version 'jdk8u172-b11'
```
Comment (commitay) [inline Casks/openjdk.rb]: Please update this so it installs the same version as the current `java` cask, JDK 10.

JDK 8 can be submitted to [`cask-versions`](https://github.com/Homebrew/homebrew-cask-versions) where we have the Oracle and Zulu JDK 8 casks.

File: Casks/openjdk.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  sha256 'c271c5aeebfec5cd6f41c24092472a66fc55f893e3118e63f4c1ab1cb4446157'
+
+  url "https://github.com/AdoptOpenJDK/openjdk8-releases/releases/download/#{version}/OpenJDK8_x64_Mac_#{version}.tar.gz"
+  appcast 'https://github.com/AdoptOpenJDK/openjdk8-releases/releases.atom'
+  name 'OpenJDK'
+  homepage 'https://github.com/AdoptOpenJDK/openjdk8-releases'
+
+  auto_updates true
+
+  binary "/Library/Java/JavaVirtualMachines/#{version}/bin/java"
```
Comment (commitay) [inline Casks/openjdk.rb]: I'm not sure about adding `java` and `javac` binaries to users PATH. 

Also, these will conflict if they are in the other OpenJDK casks which will prevent parallel installs.

**Comment by metyl** [comment]
DRY and just use https://github.com/AdoptOpenJDK/homebrew-openjdk

**Comment by joewiz** [comment]
As of Tuesday Java 11 is out, and with it a new license that prevents commercial use without a license -- effectively pushing the open source world to openjdk. The Mac build is at http://jdk.java.net/11/. We need a new "openjdk" cask that installs OpenJDK 11. We'd also benefit from version-specific casks in the caskroom-versions repo for OpenJDK 8, 9, 10.

**Comment by joewiz** [comment]
I think the openjdk cask should probably be modeled on https://github.com/Homebrew/homebrew-cask/blob/master/Casks/java.rb, no?

File: Casks/openjdk.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
+cask 'openjdk' do
+  version 'jdk8u172-b11'
+  sha256 'c271c5aeebfec5cd6f41c24092472a66fc55f893e3118e63f4c1ab1cb4446157'
+
+  url "https://github.com/AdoptOpenJDK/openjdk8-releases/releases/download/#{version}/OpenJDK8_x64_Mac_#{version}.tar.gz"
+  appcast 'https://github.com/AdoptOpenJDK/openjdk8-releases/releases.atom'
+  name 'OpenJDK'
+  homepage 'https://github.com/AdoptOpenJDK/openjdk8-releases'
+
+  auto_updates true
```
[the remark this answers is not in the document]
Comment (gjmveloso) [reply Casks/openjdk.rb]: I thought this was a built-in mechanism of Homebrew/Cask regarding evaluating appcast to automate new version changes without needing maintainers work.

I'll remove auto_updates and also change the targeted version per comment below.

Thanks,
Gustavo

File: Casks/openjdk.rb (hunk 1 above)
Comment (gjmveloso) [reply Casks/openjdk.rb]: Per my testing, removing `/Library/Java/JavaVirtualMachines/jdk-*` is still necessary to get the system in the exactly same state before initial installation. 

Reduced all Oracle-related items -- but leaving them could have negative impact, especially with parallel `java` installed.

File: Casks/openjdk.rb (hunk 3 above)
Comment (gjmveloso) [reply Casks/openjdk.rb]: Do we have a mechanism to provide precedence over other parallel installs within `brew/cask`?
Adding those binaries were the only workaround I had found to being able to use `java` after a Terminal restart.

Please advise on how to implement this better.

Thanks!

File: Casks/openjdk.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+cask 'openjdk' do
```
Comment (commitay) [inline Casks/openjdk.rb]: This token will conflict with https://github.com/Homebrew/homebrew-cask/pull/52556.

Please rename it to `adoptopenjdk`

File: Casks/openjdk.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  version '10.0.2'
+  build = '13'
```
Comment (commitay) [inline Casks/openjdk.rb]: `version '10.0.2,13'` and use version.before_comma/version.after_comma

File: Casks/openjdk.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  sha256 '5fe24ab7845005c94bddf5f28bd876b63a2a316f7ab757a5d1eddc24e229b7ae'
+
+  url "https://github.com/AdoptOpenJDK/openjdk10-releases/releases/download/jdk-#{version}+#{build}/OpenJDK10_x64_Mac_jdk-#{version}.#{build}.tar.gz"
+  appcast 'https://github.com/AdoptOpenJDK/openjdk10-releases/releases.atom'
+  name 'OpenJDK'
+  homepage 'https://github.com/AdoptOpenJDK/openjdk10-releases'
+
+  auto_updates false
```
Comment (commitay) [inline Casks/openjdk.rb]: Just remove this line, false isn't a valid key.

File: Casks/openjdk.rb [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+  
+  artifact "jdk-#{version}+#{build}/", target: "/Library/Java/JavaVirtualMachines/jdk-#{version}+#{build}"
```
Comment (commitay) [inline Casks/openjdk.rb]: This might need `sudo` to install into `/Library/Java`?

**Comment by gjmveloso** [comment]
I agree with @metyl regarding leaving AdoptOpenJDK being managed by their own tap

> DRY and just use https://github.com/AdoptOpenJDK/homebrew-openjdk

For this reason I´m changing the scope of this PR to add SapMachine as a Cask. Sounds much more reliable and up-to-date to the current latest JDK version (11).

**Review by commitay — APPROVED** [review]

**Pull request merged**